### DISCUSSION TITLE
Example for unit testing zustand+immer?

### DISCUSSION THREAD

**Discussion by isaachinman** [pmndrs/zustand#1099] [Q&A]
I'm using `zustand@4`, plus `zustand/middleware/immer` (thanks!).

My actions typically look like this:

```tsx
const { getState, setState } = useUserStore

export const fetchUser = async () => {
  try {
    setState(state => { state.user.meta = loadingMeta('loading') })

    const data = await myAPIClient.get<User>({ relativeUrl: '/v1/user' })

    setState(state => {
      state.user = { data, meta: loadingMeta('loaded') }
    })
  } catch (error: any) {
    setState(state => { state.user.meta = loadingMeta('error', error) })
  }
}
```

I'm wondering if anyone has suggestions about how to unit test such an action, given that `setState` is called with a _function_ that directly mutates the input.

The only sort of pattern I've been able to come up with so far, looks like this (pseudo code):

```tsx
describe('fetchUser', () => {
  it('calls setState with the correct payload', async () => {
    await userStoreActions.fetchUser()

    const [[call]] = (useUserStore.setState as jest.Mock).mock.calls
    const state = {
      user: {
        data: null,
      },
    }

    call(state)

    expect(state.user.data.id).toBe('test-user-id')
  })
})
```

This feels very imperfect, as it basically requires manual testing of the mutation ourselves.

Would be very grateful to hear any suggestions!

**Comment by dai-shi** [comment]
Hope someone can help.

Meanwhile, some links:
- https://github.com/pmndrs/zustand/wiki/Testing
- #968
